internal/dag/gatewayapi_processor.go

```diff
@@ -195,39 +196,115 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(route *gatewayapi_v1alpha1.HTTPRoute) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
+// computeHosts validates the hostnames for a HTTPRoute as well as validating
+// that the hostname on the HTTPRoute matches what is optionally defined on the
+// listener.hostname.
+func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname, listenerHostname *gatewayapi_v1alpha1.Hostname) ([]string, []error) {
+
 	var hosts []string
 	var errors []error
-	if len(route.Spec.Hostnames) == 0 {
+
+	// Determine the hosts on the hostnames, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the hostnames must match the Gateway hostname.
+	if len(hostnames) == 0 && listenerHostname == nil {
 		hosts = append(hosts, "*")
 		return hosts, nil
 	}
 
-	for _, host := range route.Spec.Hostnames {
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts = append(hosts, string(*listenerHostname))
+				return hosts, nil
+			}
+		}
+	}
+
+	for _, host := range hostnames {
 
 		hostname := string(host)
-		if isIP := net.ParseIP(hostname) != nil; isIP {
-			errors = append(errors, fmt.Errorf("hostname %q must be a DNS name, not an IP address", hostname))
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
```

Comment:
a `*` should only be able to come from line https://github.com/projectcontour/contour/pull/3640/files#diff-2dfb3932d84ad36571b20cb8d66d9f0b623fdbbef49a0b41eff48eaa7fa22a05R211 above, but could the order of this processing mean that we now allow `*` in a hostname on a route? (which doesn't seem like it should be allowed, per the docs)

Reply:
You're correct based on the docs. We can set a `*` for the hostname and not setting is the same thing, but agree should update.